### PR TITLE
DynamoDB model migration

### LINKED ISSUE #18: DynamoDB Migration
To support the DynamoDB migration outlined in [this proposal](https://github.com/godaddy/warehouse.ai/blob/master/docs/proposals/moving-to-dynamodb.md):

Once `warehouse-models` is [updated](https://github.com/warehouseai/warehouse-models/issues/8)

- `bffs` uses the new DynamoDB version of `warehouse-models`
- Tests are updated and use `localstack` if necessary
- Update documentation
- Published as a new major version

### PR DESCRIPTION
**Pull request by Swaagie**
## Summary

Move to dynamodb models.

- [x] requires https://github.com/warehouseai/warehouse-models/pull/12 to be merged and released
- [x] transactions replaced with retry logic in `async.series`, replaces `async.waterfall`
- [x] tests updated to use localstack
- [x] removed `createDate`. Models now have internal `createdAt`
- [x] handle connection logic and preparation of localstack/s3.
- [x] Update `cdnup` to properly create its own bucket
- [x] Use `update` rather than `create` for `buildHead`.

According to https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateItem.html `updateItem` functions as upsert and [dynamodb-x](https://github.com/baseprime/dynamodb/blob/master/README.md#updating) uses that method internally

## Changelog

Added. Fixes #18 

## Test Plan

```
npm run localstack && npm run redis
npm test
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -743,25 +664,21 @@ BFFS.prototype.key = function key(spec, wot) {
       //
       // My thought here is to create a single delimiter since if we are using
       // this as the buildId and previousBuildId, we would have to parse it in
-      // order to actually fetch the build from the build table since we are using
-      // composite keys to be nice to cassandra
+      // order to actually fetch the build from the build database table.
       //
-      return [spec.name, spec.env, spec.version, spec.locale].join('!');
```
Comment (Swaagie) [inline index.js]: should this key no longer have `spec.locale`?

File: index.js [hunk 2]
```diff
@@ -603,13 +528,11 @@ BFFS.prototype.promote = function promote(spec, callback) {
...
-    .on('data', (data) => {
-      operations.push((next) => {
-        async.parallel([
-          (cb) => BuildHead.create(this._strip(data), cb),
-          changed && ((cb) => Build.update(data, cb))
-        ].filter(Boolean), next);
-      });
+    .on('data', data => {
+      operations.push(next => async.series([
+        BuildHead.update.bind(BuildHead, this._strip(data)),
```
Comment (Swaagie) [inline index.js]: Note this is now `update` rather than `create`, should be fine see description

File: test/bffs-test.js [hunk 3]
```diff
@@ -40,32 +39,40 @@ describe('bffs', function () {
...
-      .on('error', err => console.error(err));
+    this.timeout(60000);
 
-    datastar.connect(function (err) {
-      if (err) return next(err);
+    const dynamoDriver = new DynamoDB(config.dynamodb);
+
+    dynamo.dynamoDriver(dynamoDriver);
+    models = wrhs(dynamo);
+    redis = new Redis().on('error', console.error); // eslint-disable-line no-console
```
Comment (Swaagie) [inline test/bffs-test.js]: Do we even have fuctionality that actually still stores in Redis, other than a few isolated specific tests?

Comment (jcrugzz) [reply test/bffs-test.js]: @Swaagie its mostly legacy

File: .travis.yml [hunk 4]
```diff
@@ -1,11 +1,16 @@
-dist: trusty
 sudo: required
 language: node_js
 node_js:
-- '8'
-- '10'
+  - '8'
+  - '10'
```
Comment (msluther) [inline .travis.yml]: add 12?

File: index.js (hunk 2 above)
Comment (msluther) [reply index.js]: does `update` work if its the first time a build head was created?

File: .travis.yml (hunk 4 above)
Comment (Swaagie) [reply .travis.yml]: will do, right now travis is flaky as hell tho

File: index.js (hunk 2 above)
Comment (Swaagie) [reply index.js]: See description?

File: test/bffs-test.js (hunk 3 above)
Comment (Swaagie) [reply test/bffs-test.js]: Should I just remove it now that we release a new major anyways? I think our current understanding is that builds take so long that we don't need a quick key/value store

File: index.js (hunk 2 above)
Comment (msluther) [reply index.js]: I apparently fail at reading today, both your description and your actual comment.
fail

File: test/bffs-test.js (hunk 3 above)
Comment (Swaagie) [reply test/bffs-test.js]: Made a new ticket to remove the essentially dead code

**Pull request merged**